File: src/main/java/scribbler/config/LocalConfiguration.java
package scribbler.config;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.boot.autoconfigure.web.ServerProperties;
import org.springframework.boot.context.embedded.EmbeddedServletContainerInitializedEvent;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.ApplicationListener;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.RequestEntity;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.DefaultResponseErrorHandler;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriTemplate;
import scribbler.Main;

import java.io.IOException;
import java.net.URI;
import java.util.HashMap;
import java.util.Map;

@Configuration
@ConditionalOnWebApplication
@ConditionalOnProperty(prefix = "browse", name = "enabled", havingValue = "true")
public class LocalConfiguration
{
	private static final Logger LOG = LoggerFactory.getLogger(Main.class);

	@Bean
	@ConfigurationProperties("browse")
	public BrowserProperties browserProperties()
	{
		return new BrowserProperties();
	}

	@Bean
	public EmbeddedServletContainerInitializedListener browserStartupListener()
	{
		return new EmbeddedServletContainerInitializedListener();
	}

	private static class BrowserProperties
	{
		private boolean enabled;
		private UriTemplate uri;

		public boolean isEnabled() {
			return enabled;
		}

		public void setEnabled(boolean enabled) {
			this.enabled = enabled;
		}

		public UriTemplate getUri() {
			return uri;
		}

		public void setUri(UriTemplate uri) {
			this.uri = uri;
		}
	}

	private static class EmbeddedServletContainerInitializedListener implements ApplicationListener<EmbeddedServletContainerInitializedEvent>
	{
		@Autowired
		private ServerProperties server;

		@Autowired
		private BrowserProperties browser;

		@Override
		public void onApplicationEvent(EmbeddedServletContainerInitializedEvent event)
		{
			try {
				if (server.getSsl().getTrustStore() != null) {
					System.setProperty("javax.net.ssl.trustStore", server.getSsl().getTrustStore());
				}

				Map<String, Object> variables = new HashMap<>();
				variables.put("scheme", server.getSsl().isEnabled() ? "https" : "http");
				variables.put("address", server.getAddress() == null ? "localhost" : server.getAddress());
				variables.put("port", event.getEmbeddedServletContainer().getPort());
				variables.put("context-path", server.getContextPath() == null ? "" : server.getContextPath());
				variables.put("servlet-path", server.getServletPath() == null ? "/" : server.getServletPath());
				URI uri = browser.getUri().expand(variables);

				RestTemplate template = new RestTemplate();
				template.setErrorHandler(new DefaultResponseErrorHandler() {
					@Override
					public void handleError(ClientHttpResponse response) throws IOException {
						response.getStatusCode(); // as long as we have an HTTP status code, we're good
					}
				});
				template.exchange(RequestEntity.get(uri).build(), String.class);
				boolean headless = Boolean.getBoolean("java.awt.headless");
				try {
					if (headless) {
						System.setProperty("java.awt.headless", "false");
						System.setProperty("apple.awt.UIElement", "true");
					}
					java.awt.Desktop.getDesktop().browse(uri);
				} finally {
					if (headless) {
						System.setProperty("java.awt.headless", String.valueOf(headless));
					}
				}
			} catch (Throwable e) {
				LOG.warn("Automatic browser launch aborted: " + e.toString());
			}
		}
	}
}

File: src/main/java/scribbler/config/LocalConfiguration.kt
package scribbler.config

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
import org.springframework.boot.autoconfigure.web.ServerProperties
import org.springframework.boot.context.embedded.EmbeddedServletContainerInitializedEvent
import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.context.ApplicationListener
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.http.RequestEntity
import org.springframework.http.ResponseEntity
import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.DefaultResponseErrorHandler
import org.springframework.web.client.ResourceAccessException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriTemplate
import scribbler.LOG
import java.io.IOException
import java.net.URI
import java.util.*
import kotlin.concurrent.thread

@Profile("local")
@Configuration
@ConditionalOnWebApplication
open class LocalConfiguration
{
	@Bean
	@ConfigurationProperties("browse")
	open fun browserProperties(): BrowserProperties
	{
		return BrowserProperties()
	}

	@Bean
	open fun browserStartupListener(): EmbeddedServletContainerInitializedListener
	{
		return EmbeddedServletContainerInitializedListener()
	}

	class BrowserProperties
	{
		var enabled: Boolean = false
		var uri: UriTemplate = UriTemplate("{scheme}://{address}:{port}{context-path}{servlet-path}")
	}

	class EmbeddedServletContainerInitializedListener : ApplicationListener<EmbeddedServletContainerInitializedEvent>
	{
		@Autowired
		private val server: ServerProperties? = null

		@Autowired
		private val browser: BrowserProperties? = null

		private val max = 5000L;

		override fun onApplicationEvent(event: EmbeddedServletContainerInitializedEvent)
		{
			thread(name = "startupListener") {
				if (server!!.ssl.trustStore != null) {
					System.setProperty("javax.net.ssl.trustStore", server.ssl.trustStore)
				}

				val uri = buildUri(event.embeddedServletContainer.port)
				val request = RequestEntity.get(uri).build()
				var response: ResponseEntity<String>? = null
				val start = System.currentTimeMillis()
				var exception: Throwable? = null
				var client = RestTemplate()
				client.errorHandler = object : DefaultResponseErrorHandler() {
					override fun handleError(response: ClientHttpResponse) {}
				}
				do {
					try {
						response = client.exchange(request, String::class.java)
					} catch (ex: ResourceAccessException) {
						exception = ex
					}
				} while (response?.statusCode == null && !timedOut(start, exception))

				if (response?.statusCode != null) {
					LOG.info(String.format(
							"%n----------------------------------------" +
							"%n" +
							"%n The application is ready:" +
							"%n %s" +
							"%n" +
							"%n----------------------------------------", uri
					))

					if (browser!!.enabled) {
						openBrowser(uri)
					}
				}
			}
		}

		private fun buildUri(port: Int): URI
		{
			val variables = HashMap<String, Any>()
			variables.put("scheme", if (server!!.ssl.isEnabled) "https" else "http")
			variables.put("address", if (server.address == null) "localhost" else server.address)
			variables.put("port", port)
			variables.put("context-path", if (server.contextPath == null) "" else server.contextPath)
			variables.put("servlet-path", if (server.servletPath == null) "/" else server.servletPath)
			return browser!!.uri.expand(variables)
		}

		private fun openBrowser(uri: URI)
		{
			val headless = java.lang.Boolean.getBoolean("java.awt.headless")
			try {
				if (headless) {
					System.setProperty("java.awt.headless", "false")
					System.setProperty("apple.awt.UIElement", "true")
				}
				java.awt.Desktop.getDesktop().browse(uri)
			} catch (ex: IOException) {
				LOG.debug("Failed to open desktop browser.", ex)
				LOG.error("Failed to open desktop browser: {}", ex.message);
			} finally {
				if (headless) {
					System.setProperty("java.awt.headless", headless.toString())
				}
			}
		}

		private fun timedOut(start: Long, exception: Throwable?): Boolean
		{
			if (System.currentTimeMillis() - start > max) {
				if (exception == null) {
					LOG.error("Listener timed out while waiting for server to start.")
				} else {
					LOG.error("Listener timed out while waiting for server to start: {}", exception.message)
				}
				return true;
			}
			Thread.sleep(250)
			return false;
		}
	}
}
